ACE/ace/Sock_Connect.cpp

```diff
@@ -1473,21 +1473,32 @@ ip_check (int &ipvn_enabled, int pf)
     {
 
 #if defined (ACE_WIN32)
+      static int recursing = 0;
+      if (recursing) return recursing;
+
       // as of the release of Windows 2008, even hosts that have IPv6 interfaces disabled
       // will still permit the creation of a PF_INET6 socket, thus rendering the socket
       // creation test inconsistent. The recommended solution is to get the list of
       // endpoint addresses and see if any match the desired family.
       ACE_INET_Addr *if_addrs = 0;
       size_t if_cnt = 0;
 
-      ipvn_enabled = 1; // assume enabled to avoid recursion during interface lookup.
+      // assume enabled to avoid recursion during interface lookup.
+      recursing = 1;
       ACE::get_ip_interfaces (if_cnt, if_addrs);
-      ipvn_enabled = 0;
-      for (size_t i = 0; ipvn_enabled == 0 && i < if_cnt; i++)
+      recursing = 0;
+
+      bool found = false;
+      for (size_t i = 0; !found && i < if_cnt; i++)
         {
-          ipvn_enabled = (if_addrs[i].get_type () == pf);
+          found = (if_addrs[i].get_type () == pf);
         }
       delete [] if_addrs;
+
+      // If the list of interfaces is empty, we've tried too quickly. Assume enabled, but don't cache the result
```

Comment:
Since this "assumes enabled" we might have a false positive.  What's the implication of this for the code (like INET_Addr) that needs an answer?

Reply:
For recursion, the previous code was also "assuming enabled" so hopefully this isn't a horrible thing to assume. I also don't think I ever saw anything in the code that used ipv4_enabled, only ipv6_enabled and that was only called when ACE_HAS_IPV6 is defined.

That said, it's possible this will return a different result for a Windows machine with ACE_HAS_IPV6 defined if the result is called early enough, but unlike the previous code it should eventually settle on the correct answer (rather than being stuck with the wrong one).

Reply:
I guess the question I had may really be about how this works prior to your PR.

The case where it matters is a Windows + IPv6 build.  Since this code will be called by `ACE_SOCK_Acceptor::open`, `ACE_SOCK_Dgram::open`, `ACE_INET_Addr` constructor, etc., it seems like getting the wrong answer (even temporarily) is still the wrong thing to do.

Reply:
For windows, I'm not sure how to guarantee an answer. Testing in Github Actions was definitely showing that sometimes the early calls were generating empty / bogus lists of addresses. Blocking until an answer is really available doesn't seem like a workable solution. Perhaps we could add compile-time defines for which assumptions to make, that way users could have it default one way or the other until the real answer is known? Or we could add a default run-time parameter and try to update code in the other places that use this.